--- work/usr/sbin/start_pms	2020-03-24 17:27:32.000000000 -0400
+++ work2/usr/sbin/start_pms	2020-03-26 19:25:42.232657467 -0400
@@ -24,6 +24,9 @@
 
 test -f /etc/default/plexmediaserver && . /etc/default/plexmediaserver
 
+# Activate python virtualenv
+. "${PLEX_MEDIA_SERVER_HOME}"/Resources/Python/bin/activate
+
 # Silently support PLEX_USER if provided as override
 if [ "$PLEX_USER" != "" ]; then
   export PLEX_MEDIA_SERVER_USER="$PLEX_USER"
